sql: crash with wrapped planNodes with omitted columns

I am working on a change that rewrites a plan node and I've been hitting a crash related to plan wrapping and omitted columns. I spent a lot of time on it and figured out a bunch of details but don't quite know exactly what needs to be fixed.

A minimal reproduction is to apply this diff:
```
diff --git a/pkg/sql/show_stats.go b/pkg/sql/show_stats.go
index 1ad976ca81..5caf148134 100644
--- a/pkg/sql/show_stats.go
+++ b/pkg/sql/show_stats.go
@@ -27,7 +27,7 @@ import (
 )
 
 var showTableStatsColumns = sqlbase.ResultColumns{
-	{Name: "statistics_name", Typ: types.String},
+	{Name: "statistics_name", Typ: types.String, Hidden: true},
```

and run `table [show statistics for table ab] with ordinality` which crashes:
```
panic: invalid datum type given: string, expected string[] [recovered]
	panic: panic while executing 1 statements: TABLE [SHOW STATISTICS FOR TABLE _] WITH ORDINALITY; caused by invalid datum type given: string, expected string[]

github.com/cockroachdb/cockroach/pkg/sql/sqlbase.DatumToEncDatum(0xc0015dfb88, 0x3a52020, 0xc000a3fd00, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
	/go2/src/github.com/cockroachdb/cockroach/pkg/sql/sqlbase/encoded_datum.go:172 +0x1e6
github.com/cockroachdb/cockroach/pkg/sql.(*planNodeToRowSource).Next(0xc0018a5400, 0xc00189c350, 0x3a1c560, 0xc000fec370, 0x0)
	/go2/src/github.com/cockroachdb/cockroach/pkg/sql/plan_node_to_row_source.go:183 +0x4fd
```

This is the plan node tree:
```
root@127.13.253.82:42843/defaultdb> explain (verbose) table [show statistics for table ab] with ordinality;
        tree       |  field   |    description    |                                                           columns                                                           |        ordering         
+------------------+----------+-------------------+-----------------------------------------------------------------------------------------------------------------------------+------------------------+
  render           |          |                   | (column_names, created, row_count, distinct_count, null_count, histogram_id, "ordinality")                                  | weak-key("ordinality")  
   │               | render 0 | column_names      |                                                                                                                             |                         
   │               | render 1 | created           |                                                                                                                             |                         
   │               | render 2 | row_count         |                                                                                                                             |                         
   │               | render 3 | distinct_count    |                                                                                                                             |                         
   │               | render 4 | null_count        |                                                                                                                             |                         
   │               | render 5 | histogram_id      |                                                                                                                             |                         
   │               | render 6 | "ordinality"      |                                                                                                                             |                         
   └── ordinality  |          |                   | (statistics_name[hidden,omitted], column_names, created, row_count, distinct_count, null_count, histogram_id, "ordinality") | weak-key("ordinality")  
        └── values |          |                   | (statistics_name[hidden,omitted], column_names, created, row_count, distinct_count, null_count, histogram_id)               |                         
                   | size     | 7 columns, 3 rows |                                                                                                                             |                         
```

I'm not entirely sure why the values node gets wrapped, but in any case this is the distsql plan:

![image](https://user-images.githubusercontent.com/16544120/56768363-2839e800-677c-11e9-9578-6948cbc47908.png)

The problem is that the second processor (specifically the `rowSourceToPlanNode` part) expects only non-omitted columns to be produced (and adds back the omitted columns), but the wrapped values node produces all of them. Perhaps the `planNodeToRowSource` should project away omitted columns when converting a planNode row to a stream row? But if we did that, we would break another case (`table [show zone configurations] with ordinality`):

```
root@127.156.242.172:35319/defaultdb> explain (verbose) table [show zone configurations] with ordinality;
             tree            |  field   |                       description                       |                                                      columns                                                      |        ordering         
+----------------------------+----------+---------------------------------------------------------+-------------------------------------------------------------------------------------------------------------------+------------------------+
  render                     |          |                                                         | (zone_name, config_sql, "ordinality")                                                                             | weak-key("ordinality")  
   │                         | render 0 | zone_name                                               |                                                                                                                   |                         
   │                         | render 1 | config_sql                                              |                                                                                                                   |                         
   │                         | render 2 | "ordinality"                                            |                                                                                                                   |                         
   └── ordinality            |          |                                                         | (zone_id[hidden,omitted], zone_name, cli_specifier[hidden,omitted], config_sql, "ordinality")                     | weak-key("ordinality")  
        └── render           |          |                                                         | (zone_id[hidden,omitted], zone_name, cli_specifier[hidden,omitted], config_sql)                                   | zone_name!=NULL         
             │               | render 0 | NULL                                                    |                                                                                                                   |                         
             │               | render 1 | defaultdb.crdb_internal.zones.cli_specifier             |                                                                                                                   |                         
             │               | render 2 | NULL                                                    |                                                                                                                   |                         
             │               | render 3 | defaultdb.crdb_internal.zones.config_sql                |                                                                                                                   |                         
             └── filter      |          |                                                         | (zone_id[omitted], zone_name[omitted], cli_specifier, config_yaml[omitted], config_sql, config_protobuf[omitted]) | cli_specifier!=NULL     
                  │          | filter   | defaultdb.crdb_internal.zones.cli_specifier IS NOT NULL |                                                                                                                   |                         
                  └── values |          |                                                         | (zone_id[omitted], zone_name[omitted], cli_specifier, config_yaml[omitted], config_sql, config_protobuf[omitted]) |                         
                             | size     | 6 columns, 6 rows                                       |                                                                                                                   |                         
(14 rows)
```
This has the following plan:
![image](https://user-images.githubusercontent.com/16544120/56768738-1efd4b00-677d-11e9-987d-774327415953.png)

So in this case, the post-process spec of the first processor expects that the processor internally produces omitted columns.. So maybe the problem is that we're not setting up a post-projection in the other case?

--- a/pkg/sql/distsql_physical_planner.go
+++ b/pkg/sql/distsql_physical_planner.go
@@ -297,8 +297,8 @@ var setNotSupportedError = newQueryNotSupportedError("SET / SET CLUSTER SETTING 
 // mustWrapNode returns true if a node has no DistSQL-processor equivalent.
 // This must be kept in sync with createPlanForNode.
 // TODO(jordan): refactor these to use the observer pattern to avoid duplication.
-func (dsp *DistSQLPlanner) mustWrapNode(node planNode) bool {
-	switch node.(type) {
+func (dsp *DistSQLPlanner) mustWrapNode(planCtx *PlanningCtx, node planNode) bool {
+	switch n := node.(type) {
 	case *scanNode:
 	case *indexJoinNode:
 	case *lookupJoinNode:
@@ -311,11 +311,17 @@ func (dsp *DistSQLPlanner) mustWrapNode(node planNode) bool {
 	case *limitNode:
 	case *distinctNode:
 	case *unionNode:
-	case *valuesNode:
 	case *virtualTableNode:
 	case *projectSetNode:
 	case *unaryNode:
 	case *zeroNode:
+	case *valuesNode:
+		// This is unfortunately duplicated by createPlanForNode, and must be kept
+		// in sync with its implementation.
+		if !n.specifiedInQuery || planCtx.isLocal || planCtx.noEvalSubqueries {
+			return true
+		}
+		return false
 	default:
 		return true
 	}
@@ -2363,6 +2369,9 @@ func (dsp *DistSQLPlanner) createPlanForNode(
 		// the subqueries with their results yet, which again means that we can't
 		// plan a DistSQL values node, which requires that all expressions be
 		// evaluatable.
+		//
+		// NB: If you change this conditional, you must also change it in
+		// checkSupportForNode!
 		if !n.specifiedInQuery || planCtx.isLocal || planCtx.noEvalSubqueries {
 			plan, err = dsp.wrapPlan(planCtx, n)
 		} else {
@@ -2383,10 +2392,6 @@ func (dsp *DistSQLPlanner) createPlanForNode(
 
 	default:
 		// Can't handle a node? We wrap it and continue on our way.
-		// TODO(jordan): this should only wrap the node itself, not all of its
-		// children as well. To deal with this the wrapper should use the
-		// planNode walker to retrieve all of the children of the current plan,
-		// and recurse with createPlanForNode on all of those children.
 		plan, err = dsp.wrapPlan(planCtx, n)
 	}
 
@@ -2450,7 +2455,7 @@ func (dsp *DistSQLPlanner) wrapPlan(planCtx *PlanningCtx, n planNode) (PhysicalP
 			// Continue walking until we find a node that has a DistSQL
 			// representation - that's when we'll quit the wrapping process and hand
 			// control of planning back to the DistSQL physical planner.
-			if !dsp.mustWrapNode(plan) {
+			if !dsp.mustWrapNode(planCtx, plan) {
 				firstNotWrapped = plan
 				p, err = dsp.createPlanForNode(planCtx, plan)
 				if err != nil {

